pkg/permissionclaim/permissionclaim_labeler.go

```diff
@@ -136,3 +140,105 @@ func (l *Labeler) LabelsFor(ctx context.Context, cluster logicalcluster.Name, gr
 
 	return labels, nil
 }
+
+// ObjectPermitted determines if any PermissionClaims apply to a given object's GroupResource, namespace, or name.
+// If no APIBinding or PermissionClaim exists for the GroupResource, the object is permitted.
+func (l *Labeler) ObjectPermitted(ctx context.Context, cluster logicalcluster.Name, groupResource schema.GroupResource, resourceName, resourceNamespace string) (bool, error) {
+	bindings, err := l.listAPIBindingsAcceptingClaimedGroupResource(cluster, groupResource)
+	if err != nil {
+		return true, fmt.Errorf("error listing APIBindings in %q accepting claimed group resource %q: %w", cluster, groupResource, err)
+	}
+
+	logger := klog.FromContext(ctx)
+
+	for _, binding := range bindings {
+		logger := logging.WithObject(logger, binding).WithValues(
+			"namespace", resourceNamespace, "name", resourceName,
+		)
+
+		path := logicalcluster.NewPath(binding.Spec.Reference.Export.Path)
+		if path.Empty() {
+			path = logicalcluster.From(binding).Path()
+		}
+
+		for _, claim := range binding.Spec.PermissionClaims {
+			// There is no PermissionClaim for this object, no need to compute selection
+			if claim.State != apisv1alpha1.ClaimAccepted || claim.Group != groupResource.Group || claim.Resource != groupResource.Resource {
+				continue
+			}
+			// The permission claim is relevant for this object, but the object does not match the criteria
+			if !Selects(claim, resourceName, resourceNamespace) {
+				// return if any permission claim on any binding do not allow the object.
+				logger.Info("not permitted by PermissionClaims")
+				return false, fmt.Errorf("operation not permitted")
+				continue
+			}
+		}
+	}
+	return true, nil
+}
+
+// Selects indicates whether an object's name and namespace are selected by an accepted PermissionClaim.
+func Selects(acceptableClaim apisv1alpha1.AcceptablePermissionClaim, name, namespace string) bool {
+	claim := acceptableClaim.PermissionClaim
+
+	// All and ResourceSelector are mutually exclusive. Validation should catch this, but don't leak info if it doesn't somehow.
+	if claim.All && len(claim.ResourceSelector) > 0 {
+		return false
+	}
+
+	// ResourceSelector nil check to be compatible with objects created prior to the addition of the All field
+	if claim.All || len(claim.ResourceSelector) == 0 {
+		return true
+	}
+
+	for _, selector := range claim.ResourceSelector {
+		if selectsName(selector.Names, name) && selectsNamespaces(selector.Namespaces, namespace) {
+			return true
+		}
+
+	}
+
+	return false
+}
+
+func selectsName(names []string, objectName string) bool {
+	if len(names) == 0 {
+		return true
+	}
+
+	if len(names) == 1 && names[0] == apisv1alpha1.ResourceSelectorAll {
```

Comment:
is this correct? can't you have `*` and another value according to the API ?

Reply:
If someone puts in `*, foo` should we reject that (admission) or just ignore all the individual entries because they're all covered by `*`?

Reply:
What Andy said, basically. We can certainly say that if `*` is present then all names (or namespaces) are covered, but having `*` and other entries is ambiguous.

Reply:
Let's either:

1. Allow `*` to be anywhere in the list, or
2. If `*` is present, no other items are allowed (reject with validation)

Reply:
Went with 1